<div style="border:solid black 2px; padding: 20px"><h1 style="color:black; margin-bottom:20px">Reviewer's comment v1</h1>
Halo Fawwaz !

Perkenalkan saya Dipta, disini saya akan mereview tugas Fawwaz ya.

Saya akan memberikan beberapa komen seperti dibawah ini, tolong jangan dipindah, dirubah, maupun dihapus ya.

Komen yang saya berikan akan muncul dalam warna hijau, kuning, atau pun merah seperti ini:

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

<div class="alert alert-warning">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Terdapat beberapa catatan.

</div>

<div class="alert alert-danger">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Perlu beberapa perbaikan

</div>

Dan tentu saja Fawwaz dapat menjawab saya dengan menggunakan ini

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak
</div>

# Menganalisis Risiko Gagal Bayar Peminjam

Tugas Anda adalah menyiapkan laporan untuk divisi kredit suatu bank. Anda akan mencari tahu pengaruh status perkawinan seorang nasabah dan jumlah anak yang dimilikinya terhadap probabilitas gagal bayar dalam pelunasan pinjaman. Pihak bank sudah memiliki beberapa data mengenai kelayakan kredit nasabah.

Laporan Anda akan dipertimbangkan pada saat membuat **penilaian kredit** untuk calon nasabah. **Penilaian kredit** digunakan untuk mengevaluasi kemampuan calon peminjam untuk melunasi pinjaman mereka.

[Dalam *notebook* ini, Anda dapat menemukan petunjuk dan instruksi singkat serta arahan untuk berfikir. Jangan abaikan petunjuk tersebut karena mereka dirancang untuk membekali Anda dengan struktur pengerjaan proyek dan akan membantu Anda melakukan analisis secara lebih mendalam. Sebaliknya, buatlah agar laporan ini terlihat seolah-olah dikirimkan oleh Anda kepada rekan satu tim Anda untuk menunjukkan temuan Anda - mereka tidak perlu tahu bahwa Anda mendapatkan bantuan eksternal dari kami! Untuk membantu Anda, kami menempatkan semua petunjuk yang harus Anda hapus dalam tanda kurung siku.]

[Sebelum Anda masuk ke dalam proses analisis data Anda, jelaskan tujuan proyek dan hipotesis yang akan Anda uji.]

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Buka *file* data dan baca informasi umumnya.

[Mulailah dengan mengimpor *library* dan memuat datanya. Anda mungkin akan menyadari bahwa Anda memerlukan *library* tambahan saat Anda telah mengerjakan proyek ini, dan itu merupakan hal yang normal. Hanya saja, pastikan untuk memperbarui bagian ini jika dibutuhkan.]

In [1]:
# Muat semua *library*
import pandas as pd

# Muat datanya
df = pd.read_csv('/datasets/credit_scoring_eng.csv')

## Soal 1. Eksplorasi data

**Deskripsi Data**
- `children` - jumlah anak dalam keluarga
- `days_employed` - pengalaman kerja nasabah dalam hari
- `dob_years` - usia nasabah dalam tahun
- `education` - tingkat pendidikan nasabah
- `education_id` - pengidentifikasi untuk tingkat pendidikan nasabah
- `family_status` - pengidentifikasi untuk status perkawinan nasabah
- `family_status_id` - tanda pengenal status perkawinan
- `gender` - jenis kelamin nasabah
- `income_type` - jenis pekerjaan
- `debt` - apakah nasabah pernah melakukan gagal bayar pinjaman
- `total_income` - pendapatan bulanan
- `purpose` - tujuan mendapatkan pinjaman

[Sekarang saatnya menjelajahi data kita. Anda perlu melihat berapa banyak kolom dan baris yang dimiliki oleh data, serta mencermati beberapa baris data untuk memeriksa potensi masalah dengan data.]

In [2]:
# Mari kita lihat berapa banyak baris dan kolom yang dimiliki oleh dataset kita

df.shape

(21525, 12)

In [3]:
# Mari tampilkan N baris pertama

df.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

[Jelaskan apa yang Anda lihat dan perhatikan dalam sampel data yang Anda tampilkan. Apakah terdapat masalah yang mungkin memerlukan penyelidikan dan perubahan lebih lanjut?]

Terdapat 12 kolom pada tabel.
Pada kolom days_employed terdapat nilai positif dan negatif, dan nilai angka dibelakang komanya bisa dibulatkan.
pada kolom education terdapat perbedaan huruf besar.
pada kolom total_income nilai di belakang koma dapat dibulatkan.
pada kolom purpose terdapat tujuan yang sama dengan kata- kata yang berbeda.
Kolom education dan education_id mengandungnilai yang sama begitu juga dengan family_status.


<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [4]:
# Dapatkan informasi data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
df.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

[Apakah terdapat nilai yang hilang di semua kolom atau hanya di beberapa kolom? Jelaskan secara singkat apa yang Anda amati dalam 1-2 kalimat.]

Pada kolom days_employed dan total_income terdapat 2174 data null.
Pada kolom days_employed dan total income memiliki tipe data float.

In [6]:
# Mari kita lihat tabel yang telah difilter dengan nilai yang hilang di kolom pertama yang mengandung data yang hilang
df_null = df[df['days_employed'].isnull()]
df_null

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [7]:
# Mari kita terapkan beberapa kondisi untuk memfilter data dan melihat jumlah baris dalam tabel yang telah difilter.
df_null.shape

(2174, 12)

In [8]:
df_null_persen =  df_null.shape[0] / df.shape[0]
df_null_persen
print(f'Presentasi nilai yang hilang pada kolom days_employed sebesar {df_null_persen:.1%}')

Presentasi nilai yang hilang pada kolom days_employed sebesar 10.1%


[Apakah nilai yang hilang tampak simetris? Apakah kita yakin dengan asumsi ini? Jelaskan pendapat Anda secara singkat di bagian ini. Anda mungkin ingin melakukan penyelidikan lebih lanjut, dan menghitung nilai yang hilang di semua baris dengan nilai yang hilang untuk memastikan bahwa sampel yang hilang memiliki ukuran yang sama.]

Dari tabel diatas, nilai yang hilang tampak simetris pada kolom days_employed dan total_income. Perlu dilakukan penyelidikan lebih lanjut terkait penyebaran nilai yang hilai pada setiap baris dan kolomnya

# **Kesimpulan sementara**

[Apakah jumlah baris dalam tabel yang telah difilter sesuai dengan jumlah nilai yang hilang? Kesimpulan apa yang bisa kita buat dari hal ini?]

[Hitung persentase nilai yang hilang jika dibandingkan dengan keseluruhan *dataset*. Apakah nilai yang hilang merupakan bagian data yang cukup besar? Jika demikian, Anda sebaiknya perlu mengisi nilai yang hilang. Untuk melakukannya, pertama-tama kita harus mempertimbangkan apakah data yang hilang bisa jadi disebabkan oleh karakteristik nasabah tertentu, seperti jenis pekerjaan atau yang lainnya. Anda harus memutuskan karakteristik mana yang menurut *Anda* mungkin merupakan penyebabnya. Kedua, kita harus memeriksa apakah ada ketergantungan nilai yang hilang pada nilai indikator lain dengan kolom-kolom yang mengidentifikasikan karakteristik tertentu nasabah.]

[Jelaskan langkah Anda selanjutnya dan bagaimana hubungannya dengan kesimpulan yang Anda buat sejauh ini.]

Dari data diatas, diketahui jumlah baris dalam tabel yang telah difilter sama dengan jumlah nilai yang hilang yaitu 2174 baris. 

Persentase nilai yang hilang sebesar 10.1 % dari keseluruhan dataset. jumlah tersebut mungkin cukup berpengaruh terhadap analisa yang akan dilakukan, karena itu diperlukan penyelidikan lebih lanjut terhadap pola- pola yang ada pada nilai yang hilang tersebut.

Pertama dilakukan pengamatan terhadap distribusi nilai dengan value_counts() pada setiap kolom dengan tabel yang hanya memiliki nilai yang hilang dan tabel yang tidak memiliki nilai yang hilang, kemudian mencari pola- pola yang menyebabkan adanya nilai yang hilang tersebut. 

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [9]:
# Mari kita periksa nasabah yang tidak memiliki data tentang karakteristik yang teridentifikasi dan kolom dengan nilai yang hilang
df_not_null = df[~df['days_employed'].isnull()]
df_not_null

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[19351 rows x 12 columns]

In [10]:
df_not_null.shape

(19351, 12)

In [11]:
# Periksalah distribusinya

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [13]:
df_null.describe()

children  days_employed    dob_years  education_id  \
count  2174.000000            0.0  2174.000000   2174.000000   
mean      0.552438            NaN    43.632015      0.800828   
std       1.469356            NaN    12.531481      0.530157   
min      -1.000000            NaN     0.000000      0.000000   
25%       0.000000            NaN    34.000000      0.250000   
50%       0.000000            NaN    43.000000      1.000000   
75%       1.000000            NaN    54.000000      1.000000   
max      20.000000            NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

In [14]:
df_not_null.describe()

children  days_employed     dob_years  education_id  \
count  19351.000000   19351.000000  19351.000000  19351.000000   
mean       0.537388   63046.497661     43.255336      0.819079   
std        1.371408  140827.311974     12.579170      0.550104   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      19351.000000  19351.000000   19351.000000  
mean           0.972249      0.081184   26787.568355  
std            1.420596      0.273125   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

Dari hasil statistika didapatkan data yang bersifat numerik memiliki kemiripan persebaran pada kolom yang tidak memiliki data yang hilang. Kemudian akan dilakukan pengecekan distribusi pada setiap kolom.

In [15]:
df_null['children'].value_counts()

 0     1439
 1      475
 2      204
 3       36
 20       9
 4        7
-1        3
 5        1
Name: children, dtype: int64

In [16]:
df_not_null['children'].value_counts()

 0     12710
 1      4343
 2      1851
 3       294
 20       67
-1        44
 4        34
 5         8
Name: children, dtype: int64

Pada kolom children terdapat data -1, dan 20 yang terlihat ganjil. 

In [17]:
df_null['dob_years'].value_counts()

34    69
40    66
31    65
42    65
35    64
36    63
47    59
41    59
30    58
28    57
57    56
58    56
54    55
38    54
56    54
37    53
52    53
39    51
33    51
50    51
51    50
45    50
49    50
29    50
43    50
46    48
55    48
48    46
53    44
44    44
60    39
61    38
62    38
64    37
32    37
27    36
23    36
26    35
59    34
63    29
25    23
24    21
66    20
65    20
21    18
22    17
67    16
0     10
68     9
69     5
20     5
71     5
70     3
72     2
19     1
73     1
Name: dob_years, dtype: int64

In [18]:
df_not_null['dob_years'].value_counts()

35    553
41    548
38    544
40    543
34    534
42    532
33    530
39    522
44    503
29    495
31    495
36    492
48    492
37    484
30    482
32    473
50    463
43    463
49    458
27    457
45    447
28    446
56    433
52    431
46    427
54    424
47    421
53    415
59    410
58    405
57    404
51    398
55    395
26    373
60    338
25    334
61    317
62    314
24    243
63    240
64    228
23    218
65    174
22    166
66    163
67    151
21     93
0      91
68     90
69     80
70     62
71     53
20     46
72     31
19     13
73      7
74      6
75      1
Name: dob_years, dtype: int64

Pada kolom dob_years terdapat nilai 0 yang terlihat ganjil.

In [19]:
df_null['education'].value_counts()

secondary education    1408
bachelor's degree       496
SECONDARY EDUCATION      67
Secondary Education      65
some college             55
Bachelor's Degree        25
BACHELOR'S DEGREE        23
primary education        19
Some College              7
SOME COLLEGE              7
Primary Education         1
PRIMARY EDUCATION         1
Name: education, dtype: int64

In [20]:
df_not_null['education'].value_counts()

secondary education    12342
bachelor's degree       4222
SECONDARY EDUCATION      705
Secondary Education      646
some college             613
BACHELOR'S DEGREE        251
Bachelor's Degree        243
primary education        231
Some College              40
SOME COLLEGE              22
PRIMARY EDUCATION         16
Primary Education         14
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

Pada kolom education perlu dilakukan perubahan labelling menjadi primary education, secondary education, bachelor's degree, graduate degree, dan some college.

In [21]:
df_null['education_id'].value_counts()

1    1540
0     544
2      69
3      21
Name: education_id, dtype: int64

In [22]:
df_not_null['education_id'].value_counts()

1    13693
0     4716
2      675
3      261
4        6
Name: education_id, dtype: int64

Terdapat 5 index pada kolom education_id dan pada populasi dengan data yang hilang hanya memiliki 4 index.

In [23]:
df_null['family_status'].value_counts()

married              1237
civil partnership     442
unmarried             288
divorced              112
widow / widower        95
Name: family_status, dtype: int64

In [24]:
df_not_null['family_status'].value_counts()

married              11143
civil partnership     3735
unmarried             2525
divorced              1083
widow / widower        865
Name: family_status, dtype: int64

In [25]:
df_null['family_status_id'].value_counts()

0    1237
1     442
4     288
3     112
2      95
Name: family_status_id, dtype: int64

In [26]:
df_not_null['family_status_id'].value_counts()

0    11143
1     3735
4     2525
3     1083
2      865
Name: family_status_id, dtype: int64

Pada kolom family_status dan family_status_id data yang dihasilkan terlihat sama.

In [27]:
df_null['gender'].value_counts()

F    1484
M     690
Name: gender, dtype: int64

In [28]:
df_not_null['gender'].value_counts()

F      12752
M       6598
XNA        1
Name: gender, dtype: int64

Pada kolom gender terdapat 1 nilai XNA yang mungkin bisa diubah menjadi F karena populasi F paling besar.

In [29]:
df_null['income_type'].value_counts()

employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

In [30]:
df_not_null['income_type'].value_counts()

employee                       10014
business                        4577
retiree                         3443
civil servant                   1312
unemployed                         2
student                            1
entrepreneur                       1
paternity / maternity leave        1
Name: income_type, dtype: int64

Pada kolom income_type distribusi data pada baris dengan nilai yang hilang dan yang tidak hilang terlihat sama.

In [31]:
df_null['debt'].value_counts()

0    2004
1     170
Name: debt, dtype: int64

In [32]:
df_not_null['debt'].value_counts()

0    17780
1     1571
Name: debt, dtype: int64

Pada kolom debt distribusi data pada baris dengan nilai yang hilang dan yang tidak hilang terlihat sama.

In [33]:
df_null['purpose'].value_counts()

having a wedding                            92
to have a wedding                           81
wedding ceremony                            76
construction of own property                75
housing transactions                        74
buy real estate                             72
purchase of the house for my family         71
transactions with my real estate            71
housing renovation                          70
transactions with commercial real estate    70
buy commercial real estate                  67
buying property for renting out             65
property                                    62
real estate transactions                    61
buy residential real estate                 61
housing                                     60
building a property                         59
cars                                        57
going to university                         56
to become educated                          55
second-hand car purchase                    54
buying my own

In [34]:
df_not_null['purpose'].value_counts()

wedding ceremony                            721
to have a wedding                           693
having a wedding                            685
real estate transactions                    615
buy commercial real estate                  597
purchase of the house                       595
buying property for renting out             588
housing                                     587
transactions with commercial real estate    581
building a real estate                      580
housing transactions                        579
purchase of my own house                    574
property                                    572
purchase of the house for my family         570
building a property                         561
construction of own property                560
transactions with my real estate            559
buy real estate                             552
buy residential real estate                 546
housing renovation                          542
car                                     

Pada kolom purpose distribusi data pada baris dengan nilai yang hilang dan yang tidak hilang terlihat sama. Hanya saja labelling pada kolom purpose sepertinya perlu dirapihkan.

In [35]:
# Dilakukan pengecekan antar kolom dengan kolom yang miliki nilai yang hilang

In [36]:
df_not_null.groupby('income_type')[['total_income', 'days_employed']].count()

total_income  days_employed
income_type                                             
business                             4577           4577
civil servant                        1312           1312
employee                            10014          10014
entrepreneur                            1              1
paternity / maternity leave             1              1
retiree                              3443           3443
student                                 1              1
unemployed                              2              2

Nilai pada kolom total_income dan days_employed memliki distribusi yang mirip.

In [37]:
df['days_employed'].value_counts()

-327.685916     1
-1580.622577    1
-4122.460569    1
-2828.237691    1
-2636.090517    1
               ..
-7120.517564    1
-2146.884040    1
-881.454684     1
-794.666350     1
-3382.113891    1
Name: days_employed, Length: 19351, dtype: int64

Kolom days_employed memiliki nilai negatif dan positif.

In [38]:
df_not_null[df_not_null['days_employed'] > 0]['income_type'].value_counts()

retiree       3443
unemployed       2
Name: income_type, dtype: int64

In [39]:
df_not_null[df_not_null['days_employed'] < 0]['income_type'].value_counts()

employee                       10014
business                        4577
civil servant                   1312
student                            1
entrepreneur                       1
paternity / maternity leave        1
Name: income_type, dtype: int64

Nilai positif pada kolom days employed berasal dari retiree dan unemployed.

[Deksripsikan temuan Anda di sini.]
**Kemungkinan penyebab hilangnya nilai dalam data**

[Kemukakan gagasan Anda mengenai kemungkinan alasan hilangnya nilai-nilai tersebut. Apakah menurut Anda mereka hilang secara acak atau apakah terdapat pola-pola tertentu?]

[Mari kita mulai memeriksa apakah nilai yang hilang bersifat acak.]
 
Stelah dilakukan pengecekan terhadap distribusi data tidak ditemukan pola - pola tertentu yang menjadi kemungkinan penyebab hilangnya nilai dalam data.


<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [40]:
# Memeriksa distribusi di seluruh *dataset*
df.describe()


children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

**Kesimpulan sementara**

[Apakah distribusi dalam *dataset* yang asli mirip dengan distribusi tabel yang telah difilter? Apa artinya hal tersebut untuk kita?]
Distribusi data pada dataset asli memiliki kemiripan dengan distribusi data pada tabel yang telah difilter tanpa nilai yang hilang. Nilai yang hilang pada setiap kolom memiliki distribusi yang sama. kemungkinan besar nilai yang hilang terjadi karena kebetulan atau kesalahan input.

[Jika menurut Anda kita belum dapat membuat kesimpulan apa pun, mari kita kembali menyelidiki *dataset* lebih lanjut. Mari pikirkan alasan lain yang dapat menyebabkan data hilang dan periksa apakah kita dapat menemukan pola tertentu yang dapat membuat kita berpikir bahwa hilangnya nilai-nilai tersebut tidak terjadi secara acak. Karena ini merupakan tugas Anda, maka bagian ini adalah opsional.]

In [41]:
# Periksa penyebab dan pola lain yang dapat mengakibatkan nilai yang hilang


**Kesimpulan sementara**

[Apakah pada akhirnya kita dapat memastikan bahwa nilai yang hilang adalah suatu kebetulan? Periksa hal lain yang menurut Anda penting di sini.]
Pada akhirnya tidak didapatkan pola tertentu yang menyebabkan nilai yang hilang.

In [42]:
# Periksa pola lainnya - jelaskan pola tersebut


**Kesimpulan**

[Apakah Anda menemukan suatu pola? Bagaimana Anda mendapatkan kesimpulan ini?]

[Jelaskan bagaimana Anda akan mengatasi nilai-nilai yang hilang. Pertimbangkan kategori dengan nilai yang hilang.]

[Rencanakan secara singkat langkah Anda selanjutnya untuk mentransformasi data. Anda mungkin perlu mengatasi berbagai jenis masalah: duplikat, pencatatan yang berbeda, sumber data yang salah, dan nilai yang hilang.]

Nilai yang hilang pada tabel kemungkinan terjadi karena kebetulan atau kesalahan input. karena itu, nilai yang hilang akan diisi dengan nilai mayoritas pada masing- masing kolomnya.

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Transformasi data

[Mari kita perhatikan setiap kolom untuk melihat masalah apa yang mungkin dimiliki mereka.]

[Mulailah dengan menghapus duplikat dan memperbaiki data tentang informasi pendidikan jika diperlukan.]

In [43]:
# Mari kita lihat semua nilai di kolom pendidikan untuk memeriksa ejaan apa yang perlu diperbaiki
df['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

In [44]:
# Perbaiki pencatatan jika diperlukan
df['education'] = df['education'].str.lower()



In [45]:
# Periksa semua nilai di kolom untuk memastikan bahwa kita telah memperbaikinya dengan tepat
df['education'].value_counts()


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Periksa data kolom `children`]

In [46]:
# Mari kita lihat distribusi nilai pada kolom `children`
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

[Apakah terdapat hal-hal yang aneh di kolom tersebut? Bagaimana mereka bisa terjadi? Buat keputusan tentang apa yang akan Anda lakukan dengan data ini dan jelaskan alasannya.]

Terdapat nilai yang aneh berupa -1 dan 20 yang kemungkinan besar adalah salah ketik dari 1 dan 2.

In [47]:
# [perbaiki data berdasarkan keputusan Anda]
df.loc[df['children'] == 20, ['children']] = 2
df.loc[df['children'] == -1, ['children']] = 1

In [48]:
# Periksa kembali kolom `children` untuk memastikan bahwa semuanya telah diperbaiki
df['children'].value_counts()


0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Periksa data dalam kolom the `days_employed`. Pertama-tama pikirkan tentang masalah apa yang mungkin ada pada kolom tersebut, serta pikirkan juga apa yang mungkin ingin Anda periksa dan bagaimana Anda akan melakukannya.]

In [49]:
# Temukan data yang bermasalah di kolom `days_employed` jika memang terdapat masalah, dan hitung persentasenya

In [50]:
# Kolom days_employed diubah keseluruhan nilainya menjadi positif.
df['days_employed'] = df['days_employed'].abs()

In [51]:
# Kolom days employed dikonversi kedalam satuan tahun
df['de_year'] = df['days_employed'] / 365.4

In [52]:
df['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [53]:
df['de_year'].describe()

count    19351.000000
mean       183.127337
std        380.489547
min          0.066069
25%          2.536971
50%          6.004982
75%         15.155672
max       1099.494802
Name: de_year, dtype: float64

Terdapat nilai maksimal 1099 tahun pada kolom days_employed, sehingga nilai yang tidak wajar tersebut harus di handle terlebih dahulu. Nilai 

In [54]:
df.loc[(df['days_employed'] > 320000), ['days_employed']] = ((df['dob_years'] - 17) * 365.4)

Nilai yang lebih dari 320000 hari diubah menjadi selisih umur dengan umur minimal kerja yaitu 17 tahun.

In [55]:
df[df['income_type'] == 'retiree'].groupby('income_type')['days_employed'].mean()

income_type
retiree    15396.892594
Name: days_employed, dtype: float64

Pada kategori retiree nilai rata- ratanya masa kerjanya sekitar 40++ tahun. 

In [56]:
# Terdapat nilai yang hilang pada kolom days_employed
df['days_employed'].isna().sum()

2174

In [57]:
print('Pada kolom days_employed terdapat baris dengan nilai yang hilang')
print(f'Presentasi nilai yang hilang pada kolom days_employed sebesar {df_null_persen:.1%}')

Pada kolom days_employed terdapat baris dengan nilai yang hilang
Presentasi nilai yang hilang pada kolom days_employed sebesar 10.1%


Setelah menghandle nilai yang tidak wajar, saatnya mengisi nilai yang hilang dengan mediannya.

In [58]:
# Median pada tiap kategori kolom income_type yang digunakan.
de_business = df.loc[df['income_type'] == 'business']
de_business_median = de_business['days_employed'].median()

In [59]:
de_cs = df.loc[df['income_type'] == 'civil servant']
de_cs_median = de_cs['days_employed'].median()

In [60]:
de_employee = df.loc[df['income_type'] == 'employee']
de_employee_median = de_employee['days_employed'].median()

In [61]:
de_entrepreneur = df.loc[df['income_type'] == 'entrepreneur']
de_entrepreneur_median = de_entrepreneur['days_employed'].median()

In [62]:
de_retiree = df.loc[df['income_type'] == 'retiree']
de_retiree_median = de_retiree['days_employed'].median()


In [63]:
# Atasi nilai yang bermasalah, jika ada
df.loc[(df['income_type'] == 'business'), ['days_employed']] = df.loc[(df['income_type'] == 'business'), ['days_employed']].fillna(de_business_median)

In [64]:
df.loc[(df['income_type'] == 'civil servant'), ['days_employed']] = df.loc[(df['income_type'] == 'civil servant'), ['days_employed']].fillna(de_cs_median)

In [65]:
df.loc[(df['income_type'] == 'employee'), ['days_employed']] = df.loc[(df['income_type'] == 'employee'), ['days_employed']].fillna(de_employee_median)

In [66]:
df.loc[(df['income_type'] == 'entrepreneur'), ['days_employed']] = df.loc[(df['income_type'] == 'entrepreneur'), ['days_employed']].fillna(de_entrepreneur_median)

In [67]:
df.loc[(df['income_type'] == 'retiree'), ['days_employed']] = df.loc[(df['income_type'] == 'retiree'), ['days_employed']].fillna(de_retiree_median)

In [68]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['days_employed'].isna().sum()

0

Jumlah nilai yang hilang pada kolom days_employed sudah menjadi 0. 

<div class="alert alert-warning">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

- untuk bab ini diminta jangan langsung menggunakan median ya
    - tetapi bisa menggunakan conditional fillna
    - silahkan tanya tutor/instructor
    - lalu juga masih ada nilai yg aneh yaitu masa kerja sampai 900++ tahun

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak, saya coba gunakan conditional fillna dan pisahkan nilai dengan masa kerja diatas 900++ tahun.
</div>

<div class="alert alert-danger">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

- masih ada nilai kolom days_employed yang aneh
    - nilai 900++ tahun itu tidak dipisahkan, tapi di handle, entah dihapus atau dirubah/diganti.

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak, untuk masa kerja 900++ tahun dihandle dulu dengan dibagi 100 sehingga nilainya menjadi masuk akal.
</div>

<div class="alert alert-success">
<b>Adipta's comment v.3</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Sekarang mari kita lihat usia nasabah dan mengecek apakah terdapat masalah di sana. Sekali lagi, pikirkan tentang kemungkinan kejanggalan apa yang bisa kita temui dalam kolom ini, misalnya angka usia yang tidak masuk akal.]

In [69]:
# Periksa `dob_years` untuk nilai yang mencurigakan dan hitung persentasenya
df['dob_years'].value_counts()


35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [70]:
df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

[Putuskan apa yang akan Anda lakukan dengan nilai yang bermasalah dan jelaskan alasannya.]

Ada 101 baris dengan nilai usia 0, karena nilainya tidak masuk akal jadi nilainya akan didrop.

In [71]:
# Atasi masalah pada kolom `dob_years`, jika terdapat masalah
df.drop(df[df['dob_years'] == 0].index, inplace = True)

In [72]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

<div class="alert alert-warning">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

- Nilai 0 bisa di drop saja

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak, nilai 0 nya dihapus
</div>

<div class="alert alert-success">
<b>Adipta's comment v.2</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Sekarang saatnya memeriksa kolom `family_status`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu Anda atasi.]

In [73]:
# Mari kita lihat nilai untuk kolom ini
df['family_status'].value_counts()


married              12331
civil partnership     4156
unmarried             2797
divorced              1185
widow / widower        955
Name: family_status, dtype: int64

In [74]:
# Atasi nilai yang bermasalah di `family_status`, jika ada


Dari hasil pengamatan, tidak didapatkan nilai yang bermasalah pada kolom 'family_status'.

In [75]:
# Periksa hasilnya - pastikan nilainya telah diperbaiki



[Sekarang saatnya memeriksa kolom `gender`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu Anda atasi]

In [76]:
# Mari kita liat nilai dalam kolom ini
df['gender'].value_counts()


F      14164
M       7259
XNA        1
Name: gender, dtype: int64

In [77]:
# Atasi nilai-nilai yang bermasalah, jika ada
df.loc[df['gender'] == 'XNA', ['gender']] = 'F'

Dari hasil pengamatan, terdapat 1 nilai yang tidak teridentifikasi.

In [78]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki
df['gender'].value_counts()


F    14165
M     7259
Name: gender, dtype: int64

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Sekarang saatnya memeriksa kolom `income_type`. Periksalah nilai seperti apa yang dimuat di dalam kolom ini dan masalah apa yang mungkin perlu Anda atasi]

In [79]:
# Mari kita lihat nilai dalam kolom ini
df['income_type'].value_counts()


employee                       11064
business                        5065
retiree                         3836
civil servant                   1453
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [80]:
# Atasi nilai yang bermasalah, jika ada

Dari hasil pengamatan, tidak didapatkan nilai yang bermasalah pada kolom 'income_type'.

In [81]:
# Periksa hasilnya - pastikan bahwa masalahnya telah diperbaiki


[Sekarang saatnya melihat apakah ada duplikat di dalam data kita. Jika kita menemukannya, Anda harus memutuskan apa yang akan Anda lakukan dengan duplikat tersebut dan menjelaskan alasannya.]

In [82]:
# Periksa duplikat
df.duplicated().sum()

71

In [83]:
df.shape

(21424, 13)

In [84]:
# Atasi duplikat, jika ada
df = df.drop_duplicates().reset_index(drop=True)

In [85]:
# Lakukan pemeriksaan terakhir untuk mengecek apakah kita memiliki duplikat
df.duplicated().sum()

0

In [86]:
# Periksa ukuran dataset yang sekarang Anda miliki setelah manipulasi pertama yang Anda lakukan
df.shape

(21353, 13)

[Jelaskan *dataset* baru Anda: jelaskan secara singkat apa saja perubahannya dan seberapa besar persentase perubahannya, jika ada.]


In [87]:
print('Total duplikat yang terhapus ada 71 baris sehingga jumlah baris menjadi 21454.')
print(f'presentase perubahannya sebesar {71/21525:0.1%}')

Total duplikat yang terhapus ada 71 baris sehingga jumlah baris menjadi 21454.
presentase perubahannya sebesar 0.3%


<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

# Bekerja dengan nilai yang hilang

[Untuk mempercepat pekerjaan dengan sejumlah data, Anda mungkin ingin menggunakan *dictionary* untuk beberapa nilai yang memiliki ID. Jelaskan mengapa dan *dictionary* apakah yang akan Anda gunakan.]

In [88]:
# Temukan dictionary


In [89]:
df[['education_id', 'education']].drop_duplicates().set_index('education_id')['education'].to_dict()

{0: "bachelor's degree",
 1: 'secondary education',
 2: 'some college',
 3: 'primary education',
 4: 'graduate degree'}

In [90]:
df[['family_status_id', 'family_status']].drop_duplicates().set_index('family_status_id')['family_status'].to_dict()

{0: 'married',
 1: 'civil partnership',
 2: 'widow / widower',
 3: 'divorced',
 4: 'unmarried'}

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

### Memperbaiki nilai yang hilang di `total_income`

[Jelaskan secara singkat kolom dengan nilai yang hilang manakah yang perlu Anda tangani. Jelaskan bagaimana Anda akan memperbaikinya.]


[Mulailah dengan mengatasi total nilai pendapatan yang hilang. Buatlah kategori usia untuk nasabah. Buatlah kolom baru yang memuat kategori usia. Strategi ini dapat membantu untuk menghitung total nilai pendapatan.]


In [91]:
# Mari kita tulis sebuah fungsi untuk menghitung kategori usia
def age_category(age):
    if age < 18:
        return 'teen'
    if age > 60:
        return 'old'
    return 'adult'
    

In [92]:
# Lakukan pengujian untuk melihat apakah fungsi Anda bekerja atau tidak
age_category(29)

'adult'

In [93]:
# Buatlah kolom baru berdasarkan fungsi
df['age_category'] = df['dob_years'].apply(age_category)


In [94]:
# Periksa bagaimana nilai di dalam kolom baru
df.head()


children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   13154.400000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose     de_year age_category  
0    purchase of the house   23.091607        adult  
1             car purchase   11.014789        adult  
2    purchase of the house   15.389772        adult  
3  supplementary education   11.288307        adult  
4        to have a wedding  931.215304        adult

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Pikirkan tentang faktor-faktor yang biasanya bergantung pada pendapatan. Pada akhirnya, Anda akan mengetahui apakah Anda harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Untuk membuat keputusan ini, Anda mungkin ingin melihat distribusi faktor-faktor yang Anda identifikasi sebagai dampak dari pendapatan seseorang.]

[Buatlah tabel yang hanya memuat data tanpa nilai yang hilang. Data ini akan digunakan untuk memperbaiki nilai yang hilang.]

In [95]:
# Buat tabel tanpa nilai yang hilang dan tampilkan beberapa barisnya untuk memastikan semuanya berjalan dengan baik
df_not_null_2 = df[~df['days_employed'].isnull()]
df_not_null_2.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0   13154.400000         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose     de_year age_category  
0    purchase of the house   23.091607        adult  
1             car purchase   11.014789        adult  
2    purchase of the house   15.389772        adult  
3  supplementary education   11.288307        adult  
4        to have a wedding  931.215304        adult

In [96]:
# Perhatikan nilai rata-rata untuk pendapatan berdasarkan faktor yang telah Anda identifikasi
df_not_null_2.groupby('age_category')['total_income'].mean()

age_category
adult    27207.465561
old      23057.777452
Name: total_income, dtype: float64

In [97]:
# Perhatikan nilai median untuk pendapatan berdasarkan faktor yang telah Anda identifikasi
df_not_null_2.groupby('age_category')['total_income'].median()

age_category
adult    23565.814
old      19637.056
Name: total_income, dtype: float64

[Ulangi perbandingan tersebut untuk beberapa faktor. Pastikan Anda mempertimbangkan berbagai aspek dan menjelaskan proses berpikir Anda.]



[Buatlah keputusan tentang karakteristik yang paling menentukan pendapatan dan apakah Anda akan menggunakan median atau rata-rata. Jelaskan mengapa Anda membuat keputusan ini]


In [98]:
#  Rata- rata untuk tingkat pendapatan dengan kategori usia adult
adult = df.loc[df['age_category'] == 'adult'] 
adult_median = adult['total_income'].median()

In [99]:
#  Rata- rata untuk tingkat pendapatan dengan kategori usia old
old = df.loc[df['age_category'] == 'old']
old_median = old['total_income'].median()

In [100]:
df.loc[(df['age_category'] == 'adult'), ['total_income']] = df.loc[(df['age_category'] == 'adult'), ['total_income']].fillna(adult_median)

In [101]:
df.loc[(df['age_category'] == 'old'), ['total_income']] = df.loc[(df['age_category'] == 'old'), ['total_income']].fillna(old_median)

In [102]:
df['total_income'].isnull().sum()

0

<div class="alert alert-warning">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

- nah ini yg disebut conditional fillna
    - tetapi saran saya jangan pakai mean, pakai median saja
    - kalau pakai mean itu tidak robust terhadap outliers

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak, diganti menggunakan median
</div>

<div class="alert alert-success">
<b>Adipta's comment v.2</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21353 entries, 0 to 21352
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21353 non-null  int64  
 1   days_employed     21353 non-null  float64
 2   dob_years         21353 non-null  int64  
 3   education         21353 non-null  object 
 4   education_id      21353 non-null  int64  
 5   family_status     21353 non-null  object 
 6   family_status_id  21353 non-null  int64  
 7   gender            21353 non-null  object 
 8   income_type       21353 non-null  object 
 9   debt              21353 non-null  int64  
 10  total_income      21353 non-null  float64
 11  purpose           21353 non-null  object 
 12  de_year           19260 non-null  float64
 13  age_category      21353 non-null  object 
dtypes: float64(3), int64(5), object(6)
memory usage: 2.3+ MB


[Jika Anda menemukan kesalahan dalam menyiapkan nilai untuk data yang hilang, kemungkinan ada sesuatu yang istimewa tentang data untuk kategori tersebut. Mari pikirkan - Anda mungkin ingin memperbaiki beberapa hal secara manual, jika terdapat cukup data untuk menemukan median/rata-rata.]


In [104]:
# Ganti nilai yang hilang jika terdapat kesalahan


[Setelah Anda selesai dengan `total_income`, periksa apakah jumlah total nilai di kolom ini sesuai dengan jumlah nilai di kolom lain.]

In [105]:
# Periksa jumlah entri di kolom
df['total_income'].shape


(21353,)

###  Memperbaiki nilai di `days_employed`

[Pikirkan tentang parameter yang dapat membantu Anda memperbaiki nilai yang hilang di kolom ini. Pada akhirnya, Anda akan mengetahui apakah Anda harus menggunakan nilai rata-rata atau median untuk mengganti nilai yang hilang. Anda mungkin perlu melakukan penelitian yang sama dengan yang Anda lakukan saat memperbaiki data di kolom sebelumnya.]

Karena kolom days_employed sebelumnya sudah diisi dengan median, analisa dilakukan terhadap kolom de_year yang memiliki nilai yang sama dalam satuan tahun.

In [106]:
df.groupby('age_category')[ 'de_year'].median()

age_category
adult      5.209376
old      974.797859
Name: de_year, dtype: float64

In [107]:
df.groupby('age_category')['de_year'].mean()

age_category
adult    115.391731
old      795.589915
Name: de_year, dtype: float64

In [108]:
df['de_year'].describe()

count    19260.000000
mean       183.093700
std        380.465782
min          0.066069
25%          2.536539
50%          6.013574
75%         15.160596
max       1099.494802
Name: de_year, dtype: float64

In [109]:
de_filter = df[df['de_year'] < 800]
de_filter['de_year'].describe()

count    15832.000000
mean         6.441514
std          6.307700
min          0.066069
25%          2.068002
50%          4.462757
75%          8.642028
max         50.325533
Name: de_year, dtype: float64

In [110]:
def de_level(de):
    if de < 6:
        return '5exp'
    if de < 11:
        return '11exp'
    if de < 21:
        return '20exp'    
    if de < 31:
        return '30exp'  
    if de < 41:
        return '40exp'   
    return '50exp'   
    

Karena terdapat nilai yang sangat tinggi yaitu diatas 50 tahun, makan dibuat kategori untuk rata- ratanya.

In [111]:
df['de_level'] = df['de_year'].apply(de_level)
df['de_level'].value_counts()


5exp     9620
50exp    5534
11exp    3537
20exp    2013
30exp     518
40exp     131
Name: de_level, dtype: int64

Terdapat 5548 nilai yang memiliki nilai diatas 800 tahun.

In [112]:
df.groupby('de_level')['de_year'].median()

de_level
11exp      8.000053
20exp     14.016929
30exp     24.354085
40exp     33.484276
50exp    998.865515
5exp       2.535002
Name: de_year, dtype: float64

In [113]:
df.groupby('de_level')['de_year'].mean()

de_level
11exp      8.133487
20exp     14.621492
30exp     24.927396
40exp     34.350172
50exp    995.342756
5exp       2.681576
Name: de_year, dtype: float64

[Tentukan apa yang akan Anda gunakan: rata-rata atau median. Jelaskan mengapa.]

Nilai median dan mean dari tiap kategori memiliki kemiripan, karena itu nilai median akan digunakan.

In [114]:
de_5exp = df.loc[df['de_level'] == '5exp']
de_5exp_median = de_5exp['de_year'].median()

In [115]:
de_11exp = df.loc[df['de_level'] == '11exp']
de_11exp_median = de_11exp['de_year'].median()

In [116]:
de_20exp = df.loc[df['de_level'] == '20exp']
de_20exp_median = de_20exp['de_year'].median()

In [117]:
de_30exp = df.loc[df['de_level'] == '30exp']
de_30exp_median = de_30exp['de_year'].median()

In [118]:
de_40exp = df.loc[df['de_level'] == '40exp']
de_40exp_median = de_40exp['de_year'].median()

In [119]:
de_50exp = df.loc[df['de_level'] == '50exp']
de_50exp_median = de_50exp['de_year'].median()

Setelah mendapatkan rata- rata setiap kategori, dilakukan pengubahan nilai yang hilai berdasarkan kategorinya masing- masing.

In [120]:
df.loc[(df['de_level'] == '5exp'), ['de_year']] = df.loc[(df['de_level'] == '5exp'), ['de_year']].fillna(de_5exp_median)

In [121]:
df.loc[(df['de_level'] == '11exp'), ['de_year']] = df.loc[(df['de_level'] == '11exp'), ['de_year']].fillna(de_11exp_median)

In [122]:
df.loc[(df['de_level'] == '20exp'), ['de_year']] = df.loc[(df['de_level'] == '20exp'), ['de_year']].fillna(de_20exp_median)

In [123]:
df.loc[(df['de_level'] == '30exp'), ['de_year']] = df.loc[(df['de_level'] == '30exp'), ['de_year']].fillna(de_30exp_median)

In [124]:
df.loc[(df['de_level'] == '40exp'), ['de_year']] = df.loc[(df['de_level'] == '40exp'), ['de_year']].fillna(de_40exp_median)

In [125]:
df.loc[(df['de_level'] == '50exp'), ['de_year']] = df.loc[(df['de_level'] == '50exp'), ['de_year']].fillna(de_50exp_median)

In [126]:
df['de_year'].isna().sum()

0

<div class="alert alert-warning">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

- tetapi saran saya jangan pakai mean, pakai median saja
    - kalau pakai mean itu tidak robust terhadap outliers

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
    
Baik kak, diganti menggunakan median
</div>

<div class="alert alert-success">
<b>Adipta's comment v.2</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

Jumlah nilai yang hilang = 0

In [127]:
# Periksa entri di semua kolom - pastikan kita memperbaiki semua nilai yang hilang
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21353 entries, 0 to 21352
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21353 non-null  int64  
 1   days_employed     21353 non-null  float64
 2   dob_years         21353 non-null  int64  
 3   education         21353 non-null  object 
 4   education_id      21353 non-null  int64  
 5   family_status     21353 non-null  object 
 6   family_status_id  21353 non-null  int64  
 7   gender            21353 non-null  object 
 8   income_type       21353 non-null  object 
 9   debt              21353 non-null  int64  
 10  total_income      21353 non-null  float64
 11  purpose           21353 non-null  object 
 12  de_year           21353 non-null  float64
 13  age_category      21353 non-null  object 
 14  de_level          21353 non-null  object 
dtypes: float64(3), int64(5), object(7)
memory usage: 2.4+ MB


## Pengkategorian Data

[Untuk menjawab pertanyaan dan menguji hipotesis, Anda akan bekerja dengan data yang telah dikategorikan. Lihatlah pertanyaan-pertanyaan yang diajukan kepada Anda dan yang harus Anda jawab. Pikirkan tentang data mana yang perlu dikategorikan untuk menjawab pertanyaan-pertanyaan ini. Di bawah ini, Anda akan menemukan templat yang bisa Anda gunakan untuk mengkategorikan data. Proses pertama mencakup data teks; yang kedua membahas data numerik yang perlu dikategorikan. Anda dapat menggunakan kedua petunjuk yang disarankan atau tidak sama sekali - semuanya terserah Anda.]

[Terlepas dari keputusan Anda untuk mengatasi pengkategorian, pastikan bahwa Anda secara lugas memberikan penjelasan tentang mengapa Anda membuat keputusan tersebut. Ingat: ini adalah pekerjaan Anda dan hanya Anda yang berhak membuat segala keputusan.]


In [128]:
# Tampilkan nilai data yang Anda pilih untuk pengkategorian
df['debt'].value_counts()

0    19620
1     1733
Name: debt, dtype: int64

[Mari kita memeriksa nilai unik]


In [129]:
# Periksa nilai unik
df['debt'].unique()

array([0, 1])

[Kelompok utama apakah yang dapat Anda identifikasi berdasarkan nilai uniknya?]

[Kita akan mengkategorikan data kita berdasarkan topik ini.]

Nilai unik pada kolom debt hanya 2 yaitu apakah nasabah pernah melakukan gagal bayar pinjaman atau tidak.

In [130]:
# Mari kita tulis sebuah fungsi untuk mengategorikan data berdasarkan topik umum
def debt_status(debt):
    if debt == 1:
        return 'debt'
    return 'no debt'

In [131]:
# Buat kolom yang memuat kategori dan hitung nilainya
df['debt_status'] = df['debt'].apply(debt_status)

df['debt_status'].value_counts()

no debt    19620
debt        1733
Name: debt_status, dtype: int64

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

Pada kolom purpose dilakukan pengkategorian juga karena nilai yang ada sangat bermacam- macam.

In [132]:
df['purpose'].value_counts()

wedding ceremony                            786
having a wedding                            764
to have a wedding                           760
real estate transactions                    672
buy commercial real estate                  658
buying property for renting out             649
transactions with commercial real estate    648
housing transactions                        646
housing                                     640
purchase of the house                       640
purchase of the house for my family         637
construction of own property                633
property                                    629
transactions with my real estate            627
building a real estate                      621
purchase of my own house                    619
building a property                         619
buy real estate                             618
housing renovation                          605
buy residential real estate                 603
buying my own car                       

In [133]:
df['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [134]:
def purpose_cat(name):
    if 'hous' in name or 'propert' in name or 'estat' in name or 'real' in name:
        return 'property'
    elif 'car' in name:
        return 'car'
    elif 'edu' in name or 'univers' in name:
        return 'education'
    elif 'wed' in name:
        return 'wedding'
    else:
        return 'other'

In [135]:
df['purpose_category'] = df['purpose'].apply(purpose_cat)
df['purpose_category'].value_counts()

property     10764
car           4284
education     3995
wedding       2310
Name: purpose_category, dtype: int64

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

[Jika Anda memutuskan untuk mengategorikan data numerik, Anda juga harus membuat kategori untuk data tersebut.]

Pada kolom numerik, kolom total_income dipilih untuk dikategorikan secara umum karena nilai-nilanya yang berbeda setiap baris.

In [136]:
# Lihat semua data numerik di kolom yang Anda pilih untuk pengkategorian
df['total_income'].value_counts()

23565.814    1885
19637.056     210
17312.717       2
42413.096       2
31791.384       2
             ... 
10000.392       1
99284.696       1
6264.532        1
27097.085       1
41428.916       1
Name: total_income, Length: 19257, dtype: int64

In [137]:
# Dapatkan kesimpulan statistik untuk kolomnya
df['total_income'].describe()

count     21353.000000
mean      26439.515425
std       15706.015607
min        3306.762000
25%       17224.099000
50%       23565.814000
75%       31325.325000
max      362496.645000
Name: total_income, dtype: float64

[Tentukan rentang apa yang akan Anda gunakan untuk pengelompokan dan jelaskan alasannya.]

Pada kolom total_income dilakukan pengkategorian menjadi beberapa kelompok berdasarkan rentang incomenya. 

In [138]:
# Buat fungsi yang melakukan pengkategorian menjadi kelompok numerik yang berbeda berdasarkan rentang
def income_level(income):
    if income < 10000:
        return 'low'
    if income > 30000:
        return 'high'
    return 'med'

In [139]:
# Buat kolom yang memuat kategori
df['income_level'] = df['total_income'].apply(income_level)

In [140]:
# Hitung setiap nilai kategori untuk melihat pendistribusiannya
df['income_level'].value_counts()

med     14542
high     5890
low       921
Name: income_level, dtype: int64

<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

## Memeriksa hipotesis


**Apakah terdapat korelasi antara memiliki anak dengan probabilitas melakukan gagal bayar pinjaman?**

In [141]:
# Periksa data anak dan data gagal bayar pinjaman

# Hitung persentase gagal bayar berdasarkan jumlah anak

df1 = df.pivot_table(index='children', columns='debt_status', values='debt',  aggfunc='count')
df1['percentage'] = (df1['debt'] / (df1['no debt'] + df1['debt'])).apply(lambda x: '{:.2%}'.format(x))
df1

debt_status    debt  no debt percentage
children                               
0            1058.0  12964.0      7.55%
1             442.0   4397.0      9.13%
2             202.0   1912.0      9.56%
3              27.0    301.0      8.23%
4               4.0     37.0      9.76%
5               NaN      9.0       nan%

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan yang Anda lakukan.]

Berdasarkan hasil diatas, dapat disimpulkan bahwa jumlah anak cukup berpengaruh terhadap persentase gagal bayar. Customer dengan jumlah anak 0 memiliki persentase gagal bayar paling kecil dan semakin banyak jumlah anak persentase gagal bayarnya cenderung meningkat.


**Apakah terdapat korelasi antara status keluarga dengan probabilitas melakukan gagal bayar pinjaman?**

In [142]:
# Periksa data status keluarga dan data gagal bayar pinjaman

# Hitung persentase gagal bayar berdasarkan status keluarga

df1 = df.pivot_table(index='family_status', columns='debt_status', values='debt',  aggfunc='count')
df1['percentage'] = (df1['debt'] / (df1['no debt'] + df1['debt'])).apply(lambda x: '{:.2%}'.format(x))
df1

debt_status        debt  no debt percentage
family_status                              
civil partnership   386     3744      9.35%
divorced             85     1100      7.17%
married             927    11363      7.54%
unmarried           273     2521      9.77%
widow / widower      62      892      6.50%

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan Anda.]

Berdasarkan hasil diatas, dapat disimpulkan bahwa status keluarga cukup berpengaruh terhadap persentase gagal bayar. Customer dengan status civil partnership dan unmarried memiliki persentase gagal bayar yang lebih tinggi dibanding yang lain.


**Apakah terdapat korelasi antara tingkat pendapatan dengan probabilitas melakukan gagal bayar pinjaman?**

In [143]:
# Periksa data tingkat pendapatan dan data gagal bayar pinjaman

# Hitung persentase gagal bayar berdasarkan tingkat pendapatan

df1 = df.pivot_table(index='income_level', columns='debt_status', values='debt',  aggfunc='count')
df1['percentage'] = (df1['debt'] / (df1['no debt'] + df1['debt'])).apply(lambda x: '{:.2%}'.format(x))
df1

debt_status   debt  no debt percentage
income_level                          
high           435     5455      7.39%
low             58      863      6.30%
med           1240    13302      8.53%

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan Anda.]

Berdasarkan hasil diatas, dapat disimpulkan bahwa tingkat pendapatan cukup berpengaruh terhadap persentase gagal bayar. Customer dengan penghasilan low cenderung memiliki persentase gagal bayar yang rendah sedangkan customer dengan penghasilan med memiliki persentase gagal bayar tertinggi.


**Bagaimana tujuan kredit memengaruhi persentase gagal bayar?**

In [144]:
# Periksa persentase tingkat gagal bayar untuk setiap tujuan kredit dan lakukan penganalisisan

df1 = df.pivot_table(index='purpose_category', columns='debt_status', values='debt',  aggfunc='count')
df1['percentage'] = (df1['debt'] / (df1['no debt'] + df1['debt'])).apply(lambda x: '{:.2%}'.format(x))
df1


debt_status       debt  no debt percentage
purpose_category                          
car                400     3884      9.34%
education          370     3625      9.26%
property           779     9985      7.24%
wedding            184     2126      7.97%

**Kesimpulan**

[Tulis kesimpulan Anda berdasarkan manipulasi dan pengamatan yang Anda lakukan.]

Berdasarkan hasil diatas, dapat disimpulkan bahwatujuan kredit cukup berpengaruh terhadap persentase gagal bayar. Customer dengan tujuan kredit property memiliki persentase gagal bayar terendah sedangkan customer dengan tujuan kredit car memiliki persentase gagal bayar tertinggi.



<div class="alert alert-success">
<b>Adipta's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>

# Kesimpulan umum 

[Tuliskan kesimpulan Anda di bagian akhir ini. Pastikan Anda memasukkan semua kesimpulan penting yang telah Anda buat berkaitan dengan cara Anda memproses dan menganalisis data. Kesimpulan tersebut harus membahas nilai yang hilang, duplikat, dan kemungkinan alasan serta solusi untuk data bermasalah yang harus Anda tangani.]

[Tuliskan juga kesimpulan Anda mengenai pertanyaan-pertanyaan yang diajukan di sini.]

Dari hasil analisa diatas dapat disimpulkan beberapa hal :
- Pada kategori anak persentase gagal bayar tertinggi 9.76% dari customer yang memiliki 4 anak.
- Pada kategori status keluarga customer dengan status unmarried memiliki persentase gagal bayar tertinggi 9.75%.
- Pada kategori tingkat pendapatan customer dengan tingkat pendapatan med (medium) memiliki persentase gagal bayar tertinggi 8.54%.
- Pada kategori tujuan kredit customer dengan tujuan car memiliki memiliki persentase gagal bayar tertinggi 9.36%.
- kesimpulannya semakin banyak jumlah anak, status unmarried dan tingkat pendapatan medium semakin tinggi persentase gagal bayarnya, terutama bila tujuan kreditnya adalah car.


<div class="alert alert-block alert-warning">

# General comments #1
<b>Reviewer's comment</b> <a class="tocSkip"></a> 

Good Job Fawwaz! Sebagian besar kamu telah menyelesaikan project ini dengan baik namun masih ada beberapa hal yang bisa diimprove yang saya tandai dengan comment berwarna kuning dan merah. Jika kamu telah merevisinya, projectmu akan di approve dan kamu bisa lanjut ke tahapan selanjutnya!.
    
Keep going Fawwaz, you are doing very well!
    

</div>

<div class="alert alert-block alert-warning">

# General comments #2
<b>Reviewer's comment</b> <a class="tocSkip"></a> 

Masih ada beberapa revisi
    

</div>

<div class="alert alert-block alert-success">

# General comments #3
<b>Reviewer's comment</b> <a class="tocSkip"></a> 

Good Job Fawwaz! Revisi sudah sesuai ya!
    

</div>